### PR TITLE
Interactivity polygon

### PR DESCRIPTION
**Pull request by davidmanzanares**
Fix for https://github.com/CartoDB/renderer-prototype/issues/199

### CODE REVIEW

File: src/api/interactivity.js [hunk 1]
```diff
@@ -62,9 +62,9 @@ export default class Interactivity {
     }
 
     _getFeaturesAtPosition(lngLat) {
-        const wm = projectToWebMercator({ lng: lngLat[0], lat: lngLat[1] });
+        const wm = projectToWebMercator({ lng: lngLat.lng, lat: lngLat.lat });
         const nwmc = wToR(wm.x, wm.y, { scale: WM_R, center: { x: 0, y: 0 } });
-        return this._layerList.map(layer => layer.getFeaturesAtPosition(nwmc)).reduce(Array.concat);
+        return [].concat(...this._layerList.map(layer => layer.getFeaturesAtPosition(nwmc)));
```
Comment (IagoLast) [inline src/api/interactivity.js]: utils.flatten ?

File: src/core/dataframe.js [hunk 2]
```diff
@@ -88,9 +89,52 @@ export default class Dataframe {
 
     }
 
-    _getPolygonAtPosition(pos) {
-        console.log(pos);
-        return [];
+    _getPolygonAtPosition(p) {
```
Comment (IagoLast) [inline src/core/dataframe.js]: a brief algorithm explanation will be great! (or a link)

Also maybe we could add performance tests in the future (https://github.com/speedracer/speedracer)

File: test/unit/core/dataframe.test.js [hunk 3]
```diff
@@ -13,8 +15,42 @@ describe('src/core/dataframe', () => {
...
-            it('should return a list containing the features at the given position', () => { });
+            const polygon1 = {
+                flat: [
+                    0, 0,
+                    0, 1,
+                    1, 0
+                ],
+                holes: [],
+            };
+            const df = new Dataframe({
```
Comment (IagoLast) [inline test/unit/core/dataframe.test.js]: 🌈 `dataframe`

File: src/api/interactivity.js (hunk 1 above)
Comment (davidmanzanares) [reply src/api/interactivity.js]: `[].concat(...` this is as long as `utils.flatten`, and I prefer a native form over a custom function call

File: src/core/dataframe.js (hunk 2 above)
Comment (davidmanzanares) [reply src/core/dataframe.js]: Yes, I'll explain it, and yes, performance tests in the future should be added, and this is actually very easy to benchmark

File: src/api/interactivity.js [hunk 4]
```diff
@@ -62,9 +62,9 @@ export default class Interactivity {
     }
 
     _getFeaturesAtPosition(lngLat) {
-        const wm = projectToWebMercator({ lng: lngLat[0], lat: lngLat[1] });
+        const wm = projectToWebMercator({ lng: lngLat.lng, lat: lngLat.lat });
```
Comment (IagoLast) [inline src/api/interactivity.js]: you can pass the lnglat object directly

Comment (davidmanzanares) [reply src/api/interactivity.js]: LOL Feel like a genius xD

Comment (davidmanzanares) [reply src/api/interactivity.js]: Fixed

File: test/unit/core/dataframe.test.js (hunk 3 above)
Comment (davidmanzanares) [reply test/unit/core/dataframe.test.js]: Fixed

File: src/core/dataframe.js (hunk 2 above)
Comment (davidmanzanares) [reply src/core/dataframe.js]: Fixed

File: src/core/dataframe.js [hunk 5]
```diff
@@ -162,3 +209,29 @@ export default class Dataframe {
...
+}
+
+// Tests if a point `p` is in the half plane defined by the line with points `a` and `b`
+// Returns a negative number if the result is INSIDE, returns 0 if the result is ON_LINE,
+// returns >0 if the point is OUTSIDE
+// Parameters in {x: 0, y:0} form
+function halfPlaneTest(p, a, b) {
+    // We use the cross product of `PB x AB` to get `sin(angle(PB, AB))`
+    // The result's sign is the half plane test result
+    return (p.x - b.x) * (a.y - b.y) - (a.x - b.x) * (p.y - b.y);
```
Comment (Jesus89) [inline src/core/dataframe.js]: :heart:

**Comment by Jesus89** [comment]
It seems that the `mousemove` event is launched in the `featureClick` callback :confused:

**Comment by davidmanzanares** [comment]
> It seems that the mousemove event is launched in the featureClick callback confused

It is, but that's a bug in `master`

**Comment by Jesus89** [comment]
> It is, but that's a bug in master

Yep, and there are lots of `featureClick` calls on one click. I'll fix it in master :wink:

**Pull request merged**